src/main/java/org/opensearch/ad/rest/RestExecuteAnomalyDetectorAction.java

```diff
@@ -19,73 +19,102 @@
-        this.requestTimeout = REQUEST_TIMEOUT.get(settings);
-        clusterService.getClusterSettings().addSettingsUpdateConsumer(REQUEST_TIMEOUT, it -> requestTimeout = it);
+    public RestExecuteAnomalyDetectorAction(ExtensionsRunner extensionsRunner, SDKRestClient sdkRestClient) {
+        this.settings = extensionsRunner.getEnvironmentSettings();
+        this.requestTimeout = REQUEST_TIMEOUT.get(this.settings);
+        this.sdkRestClient = sdkRestClient;
+        extensionsRunner.getSdkClusterService().getClusterSettings().addSettingsUpdateConsumer(REQUEST_TIMEOUT, it -> requestTimeout = it);
     }
 
-    @Override
     public String getName() {
         return DETECT_DATA_ACTION;
     }
 
-    @Override
-    protected RestChannelConsumer prepareRequest(RestRequest request, NodeClient client) throws IOException {
+    protected ExtensionRestResponse prepareRequest(RestRequest request) throws IOException {
         if (!EnabledSetting.isADPluginEnabled()) {
             throw new IllegalStateException(CommonErrorMessages.DISABLED_ERR_MSG);
         }
         AnomalyDetectorExecutionInput input = getAnomalyDetectorExecutionInput(request);
-        return channel -> {
-            String rawPath = request.rawPath();
-            String error = validateAdExecutionInput(input);
-            if (StringUtils.isNotBlank(error)) {
-                channel.sendResponse(new BytesRestResponse(RestStatus.BAD_REQUEST, error));
-                return;
-            }
-
-            AnomalyResultRequest getRequest = new AnomalyResultRequest(
-                input.getDetectorId(),
-                input.getPeriodStart().toEpochMilli(),
-                input.getPeriodEnd().toEpochMilli()
+
+        String rawPath = request.rawPath();
+        String error = validateAdExecutionInput(input);
+        if (StringUtils.isNotBlank(error)) {
+            return getExecuteDetectorResponse(RestStatus.BAD_REQUEST, request, null, error);
+        }
+
+        CompletableFuture<AnomalyResultResponse> futureResponse = new CompletableFuture<>();
+
+        AnomalyResultRequest anomalyResultRequest = new AnomalyResultRequest(
+            input.getDetectorId(),
+            input.getPeriodStart().toEpochMilli(),
+            input.getPeriodEnd().toEpochMilli()
+        );
+
+        sdkRestClient
+            .execute(
+                AnomalyResultAction.INSTANCE,
+                anomalyResultRequest,
+                ActionListener.wrap(r -> futureResponse.complete(r), e -> futureResponse.completeExceptionally(e))
             );
-            client.execute(AnomalyResultAction.INSTANCE, getRequest, new RestToXContentListener<>(channel));
-        };
+
+        AnomalyResultResponse anomalyResultResponse = futureResponse
+            .orTimeout(this.requestTimeout.getMillis(), TimeUnit.MILLISECONDS)
+            .join();
+
+        return getExecuteDetectorResponse(RestStatus.OK, request, anomalyResultResponse, anomalyResultResponse.toString());
```

Comment:
Creating a new method is not helping here because you are passing an extra param `anomalyResultResponse.toString()` which is not required. Passing `null` is also not a good option. I would say remove `getExecuteDetectorResponse` method and return `ExtensionRestResponse` directly from here and same for the error case above as well.

Reply:
Or if you want to keep the method pass `null` for `content` here.

Reply:
Umm, good idea. I was trying to keep consistency with other Rest actions so I kept like that. I previously had the same idea. But then thought I should play around much with the code as It should look similar with other rest actions.

Reply:
Anyways I will implement what you said above, as I completely agree with your opinion

Reply:
Okay so now I have pushed the changes.